### PR TITLE
Add titlebars to panes

### LINKED ISSUE #7290: Panes should have editable titles
# Description of the new feature/enhancement
Having a title per tab is nice, but I have run into situations where I would much rather name each pane to help with multitasking. I have not seen this feature listed on #1000 and I feel like this can be of use for #7075. Below is Terminator on linux. Double-clicking the title allows editing and the title bar can be click-dragged to move panes around.
terminator

### PR DESCRIPTION
**Pull request by HunterMitchell**
## Summary of the Pull Request
This pull request adds titlebars to panes. Below are a few features / changes:
- The pane's title can be changed by double-click or by a context menu when right clicking on the titlebar.
- The tab's title reflects the active pane's title, not the control's title.
- The pane titlebar's background color can be changed using the "tabColor" setting only right now.
- If no "tabColor" is chosen, the default is `ChromeDisabledHigh`.
- The pane's foreground color is dynamically changed depending on the brightness of its background.

 
## References
This pull request should close #7290 and provides #7075 with a solution. 

## PR Checklist
* [X] Closes #7290 
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [X] Schema updated.
* [X] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #7290 

### Current Issues with this PR
- A crash occurs when closing a pane child and then attempting to edit the pane title. I am looking for some guidance here, as I could not figure out the root cause of the issue. Inside the `_CloseChild` method, if I don't release `_firstChild` or `_secondChild`, it works (however creating a memory leak). 
- Currently, setting the tab's background color with the color picker does not update the pane titlebar color. 
- Code Quality – I am sure there are better ways of doing some of the things in this PR (new to the codebase and C++ isn't my primary language). I would be more than happy to make edits based on feedback received. 

### Future Thoughts:
- Add support for changing the pane titlebar background/foreground colors, font/font-size, etc. directly. This was discussed here: https://github.com/microsoft/terminal/issues/7290#issuecomment-674887400
- Add support for changing the pane titlebar background/foreground colors to indicate if a pane is active. My thoughts on that start here: https://github.com/microsoft/terminal/issues/7290#issuecomment-678266263

## Validation Steps Performed
1. Set `showPaneTitlebar` to `true` if not already.
2. Double-Click the title or use the context-menu by right-clicking the titlebar to edit the title.
3. Confirm by pressing `enter` or clicking away from textbox. Press `escape` to cancel the edit.

### CODE REVIEW

**Comment by zadjii-msft** [comment]
For those curious for a preview:
image

image
Note that my ubuntu profile has acrylic pretty agressively enabled. What is important:
* the pane's title color reflects the control "tab color" (or title color). 
* The tab reflects the focused pane's "title color" and "title text". So selecting the "foo" pane changes the tab text to "foo", and the tab color to purple, while selecting the "cmd" pane changes the tab color to the default black, and the tab text to "cmd".
* this looks awesome with focus mode:
image
* Maybe the default pane titlebar color should be the same as the focused tab item background? That might lead to more consistency with the UX when a pane has a title color

<hr>

re: 
> Currently, setting the tab's background color with the color picker does not update the pane titlebar color.

I think that's fine. Currently, the order of precedence is [tab runtime color, control title color, theme color (unimplemented), default color]. So I could imagine this going two ways:
* convert the "tab runtime color" to be a "pane runtime color". Then, using the color picker on the tab would set the active _pane_'s runtime color. This comes with the detrimental side effect that you'd have to re-color each pane in a tab manually.
* add an additional "pane runtime color", and have the tab's color be composed from [tab runtime color, pane color (which is itself [pane runtime color, control title color, theme pane title color]), theme color, default]. This would allow panes to have their color changed at runtime with a color picker, and also have the tab have a manual color override. 

I'm of course just hypothesizing here in the comments - I don't think any of this is something that _needs_ to get done in _this_ PR. We can always loop back on the issue in a follow-up.

File: src/cascadia/TerminalApp/GlobalAppSettings.h [hunk 1]
```diff
@@ -82,6 +82,9 @@ class TerminalApp::GlobalAppSettings final
     GETSET_PROPERTY(bool, StartOnUserLogin, false);
     GETSET_PROPERTY(bool, AlwaysOnTop, false);
 
+    // PaneTitlebar Settings
+    GETSET_PROPERTY(bool, ShowPaneTitlebar, true);
```
[the remark this answers is not in the document]
Comment (HunterMitchell) [reply src/cascadia/TerminalApp/GlobalAppSettings.h:86]: Agreed. I think I got a little carried away 😂.

File: src/cascadia/TerminalApp/Pane.cpp [hunk 2]
```diff
@@ -55,6 +58,283 @@ Pane::Pane(const GUID& profile, const TermControl& control, const bool lastFocus
...
+    // Append Custom Grid to the root
+    _border.Child(_paneTitlebarGrid);
+
+    _UpdatePaneTitle();
+    _CreateContextMenu();
+
+    _root.Children().Append(_border);
+
+    // Update the pane title to reflect the term title
+    _control.TitleChanged([this](winrt::hstring newTitle) {
```
[the remark this answers is not in the document]
Comment (HunterMitchell) [reply src/cascadia/TerminalApp/Pane.cpp:109]: Attempting to get a weak reference to `this` results in a runtime crash. The function containing this code is called only in the Pane's constructor.

File: src/cascadia/TerminalApp/Pane.cpp [hunk 3]
```diff
@@ -55,6 +58,283 @@ Pane::Pane(const GUID& profile, const TermControl& control, const bool lastFocus
...
+    paneTextBox.MinHeight(0);
+    paneTextBox.Padding(ThicknessHelper::FromUniformLength(0));
+    paneTextBox.Margin(ThicknessHelper::FromUniformLength(0));
+    paneTextBox.HorizontalAlignment(HorizontalAlignment::Stretch);
+    paneTextBox.VerticalAlignment(VerticalAlignment::Stretch);
+    paneTextBox.TextAlignment(TextAlignment::Center);
+    paneTextBox.Height(_paneTitlebarTitle.ActualHeight());
+
+    // Setup events
+    auto weakThis{ shared_from_this() };
```
[the remark this answers is not in the document]
Comment (HunterMitchell) [reply src/cascadia/TerminalApp/Pane.cpp:145]: Went ahead and fixed this, but unfortunately, does not seem to bee the solution to the crash.

File: src/cascadia/TerminalApp/Pane.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Pane.cpp:109]: Oh well that's annoying. We might need to do add a TODO to follow-up after #3999 to make those take weakrefs. I'm not sure how we could fix those at the moment. Maybe more coffee will help...

File: src/cascadia/TerminalApp/GlobalAppSettings.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/GlobalAppSettings.h:86]: Don't worry about it - judging by the team's reaction in our sync yesterday, they might fight me on this haha

File: src/cascadia/TerminalApp/Pane.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (HunterMitchell) [reply src/cascadia/TerminalApp/Pane.cpp:109]: Sounds good, let me know what else I can do to help! Also, please note the issue regarding a crash after a child pane is closed still exists even though I marked that comment as being resolved. ☕

**Comment by HunterMitchell** [comment]
So, here's a quick update. I fixed some of the comments above, however,  the issue regarding a crash after a child pane is closed still exists and I have not been able to find a fix in my free time. This maybe held it by #3999?

**Comment by HunterMitchell** [comment]
Just wanted to get this back into the spotlight in case it got lost in the pile of other feature requests / issues. I know you showed some interested in this feature @DHowett, so would you mind also taking a look at some of the code here and maybe provide some insight. I didn't want to bother @zadjii-msft since he is out. Let me know what you think!

**Review by DHowett — COMMENTED** [review]
I love this, and I think it's the right thing for us to do.

I've got a couple minor concerns that might amount to a lot of work, unfortunately. We're not really in a good place to keep adding UI to Pane. It's my goal to eventually get us to a `.xaml` file that we can move all the control code into leaving `Pane.cpp` to host the logic. That'll require us to do a bit of work to _make `Pane` a XAML control itself_, which I'm certain is the correct thing to do.

I think a titlebar needs to be a part of that. I'm undecided as to whether this should be blocked by that work.

Thanks so much for writing this up. It's really promising, and I love how it looks!

**Comment by HunterMitchell** [comment]
As much as I want this feature done as quickly as possible, I would rather us do the harder work up front so additions to the Pane titlebar and Pane itself are easier. In this case I agree that this PR should blocked until there is a better solution like you provided. I would be up for helping rework some of it, but I have been busy recently.

**Comment by DHowett** [comment]
> In this case I agree that this PR should be blocked until there is a better solution like you provided. I would be up for helping rework some of it, but I have been busy recently.

I'll close this for now, but not as a rejection. This is a sound idea that we'll revisit when we can support it best :smile:

Thanks so much for doing the lifting here.

**Pull request closed without merging**